packages/panzer/disc-fe/src/evaluators/Panzer_Integrator_CurlBasisDotVector_impl.hpp

```diff
@@ -40,424 +40,776 @@
-	                                  p.get< Teuchos::RCP<panzer::IntegrationRule> >("IR")->dl_vector );
-     this->addDependentField(flux_vector);
-  }
-  else {
-     flux_scalar = PHX::MDField<const ScalarT,Cell,IP>( p.get<std::string>("Value Name"), 
-   	                                  p.get< Teuchos::RCP<panzer::IntegrationRule> >("IR")->dl_scalar );
-     this->addDependentField(flux_scalar);
-  }
-
-  this->addEvaluatedField(residual);
-  
-  multiplier = p.get<double>("Multiplier");
-  if (p.isType<Teuchos::RCP<const std::vector<std::string> > >("Field Multipliers")) 
+  /////////////////////////////////////////////////////////////////////////////
+  //
+  //  Main Constructor
+  //
+  /////////////////////////////////////////////////////////////////////////////
+  template<typename EvalT, typename Traits>
+  Integrator_CurlBasisDotVector<EvalT, Traits>::
+  Integrator_CurlBasisDotVector(
+    const panzer::EvaluatorStyle&   evalStyle,
+    const std::string&              resName,
+    const std::string&              valName,
+    const panzer::BasisIRLayout&    basis,
+    const panzer::IntegrationRule&  ir,
+    const double&                   multiplier, /* = 1 */
+    const std::vector<std::string>& fmNames     /* =
+      std::vector<std::string>() */)
+    :
+    evalStyle_(evalStyle),
+    useDescriptors_(false),
+    multiplier_(multiplier),
+    basisName_(basis.name())
   {
-    const std::vector<std::string>& field_multiplier_names = 
-      *(p.get<Teuchos::RCP<const std::vector<std::string> > >("Field Multipliers"));
-
-    for (std::vector<std::string>::const_iterator name = field_multiplier_names.begin(); 
-      name != field_multiplier_names.end(); ++name) 
+    using Kokkos::View;
+    using panzer::BASIS;
+    using panzer::Cell;
+    using panzer::Dim;
+    using panzer::EvaluatorStyle;
+    using panzer::IP;
+    using panzer::PureBasis;
+    using PHX::MDField;
+    using std::invalid_argument;
+    using std::logic_error;
+    using std::string;
+    using Teuchos::RCP;
+
+    // Ensure the input makes sense.
+    TEUCHOS_TEST_FOR_EXCEPTION(resName == "", invalid_argument, "Error:  "    \
+      "Integrator_CurlBasisDotVector called with an empty residual name.")
+    TEUCHOS_TEST_FOR_EXCEPTION(valName == "", invalid_argument, "Error:  "    \
+      "Integrator_CurlBasisDotVector called with an empty value name.")
+    RCP<const PureBasis> tmpBasis = basis.getBasis();
+    TEUCHOS_TEST_FOR_EXCEPTION(not tmpBasis->isVectorBasis(), logic_error,
+      "Error:  Integrator_CurlBasisDotVector:  Basis of type \""
+      << tmpBasis->name() << "\" is not a vector basis.");
+    TEUCHOS_TEST_FOR_EXCEPTION(not tmpBasis->requiresOrientations(),
+      logic_error, "Error:  Integrator_CurlBasisDotVector:  Basis of type \""
+      << tmpBasis->name() << "\" does not require orientations.  This seems " \
+      "very strange, so I'm failing.");
```

Comment:
Could we make a more exact statement than "This seems very strange so I'm failing"? Maybe say something like "Throwing since vector bases should require orientation!"?

Reply:
Sounds good to me.